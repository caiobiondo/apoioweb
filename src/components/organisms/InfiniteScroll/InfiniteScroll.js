import React, { Component } from 'react';
import propTypes from 'prop-types';
import { Loading, CircularProgress } from 'natura-ui';
import ReactInfiniteScroller from 'react-infinite-scroller';

import { debounce } from 'utils/debounce';

import { LoadingWrapper } from './InfiniteScroll.styles';

export class InfiniteScroll extends Component {
  isEmpty = () => {
    return !this.props.items || !this.props.items.length;
  };

  onScroll = debounce(() => {
    this.props.onScroll();
  }, this.props.debounce);

  render() {
    const { loading, hasMore, children, items } = this.props;

    if (!items && loading) {
      return <Loading background="transparent" />;
    }

    if (this.isEmpty()) {
      return this.props.emptyList;
    }

    return (
      <ReactInfiniteScroller
        loadMore={this.onScroll}
        hasMore={hasMore}
        loader={
          <LoadingWrapper>
            <CircularProgress thickness={2} />
          </LoadingWrapper>
        }
      >
        {children}
      </ReactInfiniteScroller>
    );
  }
}

InfiniteScroll.propTypes = {
  onScroll: propTypes.func.isRequired,
  hasMore: propTypes.bool,
  loading: propTypes.bool.isRequired,
  children: propTypes.node.isRequired,
  debounce: propTypes.number,
  items: propTypes.array,
};

InfiniteScroll.defaultProps = {
  debounce: 0,
  hasMore: false,
  items: null,
};

export default InfiniteScroll;
