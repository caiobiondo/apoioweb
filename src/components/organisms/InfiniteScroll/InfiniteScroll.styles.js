import { LoadingWrapperStyle } from 'styles/mixins';

export const LoadingWrapper = LoadingWrapperStyle;

export const fullContainer = {
  height: '100%',
};
