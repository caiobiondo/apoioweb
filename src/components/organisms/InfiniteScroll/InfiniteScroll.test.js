import React from 'react';
import { shallow } from 'enzyme';
import { InfiniteScroll } from './InfiniteScroll';

jest.mock('utils/debounce', () => ({
  debounce: (method, time) => () => {
    method();
  },
}));

const setup = propOverrides => {
  const props = Object.assign(
    {
      onScroll: jest.fn(),
      hasMore: false,
      loading: false,
      debounce: 0,
      items: [],
    },
    propOverrides,
  );

  const result = shallow(
    <InfiniteScroll {...props}>
      <div>Test</div>
    </InfiniteScroll>,
  );

  return {
    props,
    result,
  };
};

describe('Infinite Scroll', () => {
  it('renders correctly when it is loading and there is no items', () => {
    // given
    // when
    const { result } = setup({ loading: true, items: null });

    // then
    expect(result).toMatchSnapshot();
  });

  it('renders correctly when it is an empty list', () => {
    // given
    // when
    const { result } = setup({ items: [] });

    // then
    expect(result).toMatchSnapshot();
  });

  it('renders correctly when there are items on the list', () => {
    // given
    // when
    const { result } = setup({ items: [{ id: 1 }] });

    // then
    expect(result).toMatchSnapshot();
  });

  it('should call a debounced method when onScroll is called', () => {
    // given
    const onScroll = jest.fn();
    // when
    const { result } = setup({ items: [], onScroll });
    const instance = result.instance();
    instance.onScroll();

    // then
    expect(onScroll).toBeCalled();
  });
});
