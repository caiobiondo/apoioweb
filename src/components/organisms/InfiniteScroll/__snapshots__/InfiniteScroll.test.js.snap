// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Infinite Scroll renders correctly when it is an empty list 1`] = `""`;

exports[`Infinite Scroll renders correctly when it is loading and there is no items 1`] = `
<Loading
  background="transparent"
  visible={true}
  zIndex={9999}
/>
`;

exports[`Infinite Scroll renders correctly when there are items on the list 1`] = `
<InfiniteScroll
  element="div"
  hasMore={false}
  initialLoad={true}
  isReverse={false}
  loadMore={[Function]}
  loader={
    <mixins__LoadingWrapperStyle>
      <CircularProgress
        max={100}
        min={0}
        mode="indeterminate"
        size={40}
        thickness={2}
        value={0}
      />
    </mixins__LoadingWrapperStyle>
  }
  pageStart={0}
  ref={null}
  threshold={250}
  useCapture={false}
  useWindow={true}
>
  <div>
    Test
  </div>
</InfiniteScroll>
`;
